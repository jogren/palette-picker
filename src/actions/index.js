export const setCurrentPalette = (colors) => ({
  type: 'SET_PALETTE',
  colors
});

export const setCurrentProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const setSelectedPalettes = (palettes) => ({
  type: 'SELECTED_PALETTES',
  palettes
});

export const setCurrentPaletteId = (id, name) => ({
  type: 'SET_PALETTE_ID',
  id,
  name
});