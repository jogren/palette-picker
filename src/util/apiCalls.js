export const postNewProject = async (name) => {
  const url = 'https://palette-picker-api-sfjo.herokuapp.com/api/v1/projects';
  let options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ name })
  }

  const response = await fetch(url, options);
  if(!response.ok) {
    throw Error('There was an issue posting your project');
  }

  const project = await response.json();
  console.log(project);
  return project;
}