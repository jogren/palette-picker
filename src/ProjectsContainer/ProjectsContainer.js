import React, { Component } from 'react';
import { postNewProject, getAllProjects } from '../util/apiCalls';
import { setCurrentProjects } from '../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

class ProjectsContainer extends Component {
  constructor() {
    super();
    this.state = {
      name: ""
    }
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value })
  }

  handleSubmit = async (e) => {
    const { name } = this.state;
    const { setCurrentProjects } = this.props;
    e.preventDefault();
    await postNewProject(name);
    const projects = await getAllProjects();
    console.log(projects)
    setCurrentProjects(projects)
    this.setState({ name: "" });

  }

  render() {
    const { currentProjects } = this.props;
    let projectList = currentProjects.map(project => {
      return <button>{project.name}</button>
    })
    return (
      <section>
        <form>
          <input 
            type="text"
            placeholder="New Project's Name"
            name="name"
            value={this.state.name}
            onChange={this.handleChange} />
          <button onClick={this.handleSubmit}>Submit Project Name</button>
        </form>
        <div className="div_project-name-container">
          {projectList}
        </div>
      </section>
    )
  }
}

const mapStateToProps = ({ currentProjects }) => ({
  currentProjects
})

const mapDispatchToProps = dispatch => (
  bindActionCreators({ setCurrentProjects }, dispatch)
)
export default connect(mapStateToProps, mapDispatchToProps)(ProjectsContainer);