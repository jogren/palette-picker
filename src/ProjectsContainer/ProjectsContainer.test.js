import React from 'react';
import { shallow } from 'enzyme';
import { ProjectsContainer } from './ProjectsContainer';
import { setCurrentProjects, setSelectedPalettes, hasErrored } from '../actions';
import { postNewProject, getAllProjects, getSelectedPalettes, deleteProjectFromDB } from '../util/apiCalls';
import { mapStateToProps, mapDispatchToProps } from './ProjectsContainer';


jest.mock('../util/apiCalls');

describe('ProjectsContainer', () => {
  let wrapper;
  const setSelectedPalettesMock = jest.fn();
  const setCurrentProjectsMock = jest.fn();
  const hasErroredMock = jest.fn();
  const setCurrentProjectIdMock = jest.fn();
  const currentProjectsMock = [{
    id: 3,
    name: "Winter",
    created_at: "2019-10-10T19:13:08.873Z",
    updated_at: "2019-10-10T19:13:08.873Z",
  }];

  beforeEach(() => {
    wrapper = shallow(<ProjectsContainer
      setCurrentProjects={setCurrentProjectsMock}
      setSelectedPalettes={setSelectedPalettesMock}
      currentProjects={currentProjectsMock}
      hasErrored={hasErroredMock}
      setCurrentProjectId={setCurrentProjectIdMock}
      />)
  });
  
  it('should match the snapshot with the data passed through', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update name and hasErrored state onChange of input', () => {
    wrapper.find('input').at(0).simulate('change', { target: { value: 'New Project', name: 'name' } });
    expect(wrapper.state('name')).toEqual('New Project');
    expect(wrapper.state('hasErrored')).toEqual("");
  });

  it('should invoke postNewProject, getAllProjects and setCurrentProjects when promise resolves', async () => {
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.instance().handleSubmit(mockEvent);

    await expect(postNewProject).toHaveBeenCalled();
    await expect(getAllProjects).toHaveBeenCalled();
    expect(setCurrentProjectsMock).toHaveBeenCalled();
    expect(wrapper.state('name')).toEqual("");
  });

  it('should update state if promise rejects after handleSubmit call', async () => {
    const mockEvent = { preventDefault: jest.fn() }

    postNewProject.mockImplementation(() => {
      throw Error('new error')
    })
    await wrapper.instance().handleSubmit(mockEvent);

    expect(wrapper.state('name')).toEqual("");
    expect(wrapper.state('hasErrored')).toEqual("new error");
  });

  it('should call getSelectedPalettes when handleProjectSelect is invoked and promise resolves', async () => {
    wrapper.instance().handleProjectSelect(3);

    expect(setCurrentProjectIdMock).toHaveBeenCalledWith(3);
    await expect(getSelectedPalettes).toHaveBeenCalledWith(3);
  });

  it('should call hasErrored when handleProjectSelect is called and a promise rejects', async () => {
    getSelectedPalettes.mockImplementation(() => {
      throw Error('new error')
    })
    await wrapper.instance().handleProjectSelect(1);

    expect(hasErroredMock).toHaveBeenCalledWith('new error')
  });

  it('should invoke deleteProjectFromDB, getAllProjects and setCurrentProjects when deleteProject is invoked', async () => {
    wrapper.instance().deleteProject(3);

    await expect(deleteProjectFromDB).toHaveBeenCalledWith(3);
    await expect(getAllProjects).toHaveBeenCalled();
    expect(setCurrentProjectsMock).toHaveBeenCalled();
  });
});

describe('mapDispatchToProps', () => {
  it('should call dispatch with an setCurrentProjects action', () => {
    // Setup
    const mockDispatch = jest.fn();
    const actionToDispatch = setCurrentProjects([{ name: 'palette'}]);

    // Execution
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setCurrentProjects([{ name: 'palette' }]);

    // Expectaion
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('should call dispatch with an setSelectedPalettes action', () => {
    // Setup
    const mockDispatch = jest.fn();
    const actionToDispatch = setSelectedPalettes({ name: 'palette' });

    // Execution
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setSelectedPalettes({ name: 'palette' });

    // Expectaion
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('should call dispatch with an hasErrored action', () => {
    // Setup
    const mockDispatch = jest.fn();
    const actionToDispatch = hasErrored('There was an error');

    // Execution
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.hasErrored('There was an error');

    // Expectaion
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});

describe('mapStateToProps', () => {
  it('should return an array with the currentProjects', () => {
    // Setup
    const mockState = {
      currentProjects: [{ name: 'Project 1', id: 1 }],
      currentPalette: ['color1', 'color2', 'color3', 'color4', 'color5'],
      selectedPalettes: ['selectedPalettes'],
      currentPaletteId: null,
      errorMsg: "",
      
    };
    const expected = {
      currentProjects: [{ name: 'Project 1', id: 1 }],
    };

    // Execution
    const mappedProps = mapStateToProps(mockState);

    // Expectation
    expect(mappedProps).toEqual(expected);
  });
});