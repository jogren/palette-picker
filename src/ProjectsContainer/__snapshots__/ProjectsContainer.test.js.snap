// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsContainer should match the snapshot with the data passed through 1`] = `
<section
  className="ProjectsContainer_section"
>
  <form>
    <label
      htmlFor="create-project"
    >
      Create a Project
    </label>
    <input
      autoComplete="off"
      id="create-project"
      name="name"
      onChange={[Function]}
      placeholder="Enter Name..."
      type="text"
      value=""
    />
    <button
      disabled={true}
      onClick={[Function]}
    >
      Submit Project Name
    </button>
  </form>
  <div
    className="error"
  />
  <div
    className="div_project-name-container"
  >
    <div
      className="div_buttons-container"
      key="0"
    >
      <button
        className="project-name"
        id="button_project-name"
        onClick={[Function]}
      >
        Winter
      </button>
      <button
        className="trash-btn"
        onClick={[Function]}
      >
        <FiTrash2
          className="trash-img-project"
        />
      </button>
    </div>
  </div>
</section>
`;
